libnestutil/iaf_propagator.cpp

```diff
@@ -0,0 +1,132 @@
+/*
+ *  iaf_propagator.cpp
+ *
+ *  This file is part of NEST.
+ *
+ *  Copyright (C) 2004 The NEST Initiative
+ *
+ *  NEST is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  NEST is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "iaf_propagator.h"
+
+
+IAFPropagator::IAFPropagator()
+  : tau_syn_( 0.0 )
+  , tau_m_( 0.0 )
+  , c_m_( 0.0 )
+  , alpha_( 0.0 )
+  , beta_( 0.0 )
+  , gamma_( 0.0 )
+{
+}
+
+IAFPropagator::IAFPropagator( double tau_syn, double tau_m, double c_m )
+  : tau_syn_( tau_syn )
+  , tau_m_( tau_m )
+  , c_m_( c_m )
+  , alpha_( 1 / ( c_m * tau_m * tau_m ) * ( tau_syn - tau_m ) )
+  , beta_( tau_syn * tau_m / ( tau_m - tau_syn ) )
+  , gamma_( beta_ / c_m )
+{
+}
+
+std::tuple< double, double, double, double >
+IAFPropagator::evaluate_P32_( double h ) const
+{
+  const double exp_h_tau_syn = std::exp( -h / tau_syn_ );
+  const double expm1_h_tau = numerics::expm1( -h / tau_m_ + h / tau_syn_ );
+
+  double P32 = gamma_ * exp_h_tau_syn * expm1_h_tau;
+
+  double exp_h_tau = 0.0; // exp_h_tau is only relevant if condition below fulfilled
+  if ( std::abs( tau_m_ - tau_syn_ ) < 0.1 )
+  {
+    exp_h_tau = std::exp( -h / tau_m_ );
+
+    const double P32_singular = h / c_m_ * exp_h_tau;
+    if ( tau_m_ == tau_syn_ )
+    {
+      P32 = P32_singular;
+    }
+    else
+    {
+      const double P32_linear = alpha_ * h * h * exp_h_tau / 2.;
+      const double dev_P32 = std::abs( P32 - P32_singular );
+
+      if ( dev_P32 > 2 * std::abs( P32_linear ) )
+      {
+        P32 = P32_singular;
+      }
+    }
+  }
+
+  return std::make_tuple( P32, exp_h_tau_syn, expm1_h_tau, exp_h_tau );
+}
+
+IAFPropagatorExp::IAFPropagatorExp()
+  : IAFPropagator( 0., 0., 0. )
+{
+}
+
+IAFPropagatorExp::IAFPropagatorExp( double tau_syn, double tau_m, double c_m )
+  : IAFPropagator( tau_syn, tau_m, c_m )
+{
+}
+
+
+IAFPropagatorAlpha::IAFPropagatorAlpha()
+  : IAFPropagator( 0., 0., 0. )
+{
+}
+
+IAFPropagatorAlpha::IAFPropagatorAlpha( double tau_syn, double tau_m, double c_m )
+  : IAFPropagator( tau_syn, tau_m, c_m )
+{
+}
+
+std::tuple< double, double >
+IAFPropagatorAlpha::evaluate( double h ) const
+{
+  double P32;
+  double exp_h_tau_syn;
+  double expm1_h_tau;
+  double exp_h_tau;
+  std::tie( P32, exp_h_tau_syn, expm1_h_tau, exp_h_tau ) = evaluate_P32_( h );
+
+  double P31 = gamma_ * exp_h_tau_syn * ( beta_ * expm1_h_tau - h );
+
+  if ( std::abs( tau_m_ - tau_syn_ ) < 0.1 )
```

Comment:
Could you make this 0.1 a ``const static double`` member of the class and add it to the class docstring?

Reply:
I now moved this test into a method and added a comment pointing out that the precise value of the limit plays little role.